data/blog/2024-10-18-react-server-components.mdx

```diff
@@ -0,0 +1,230 @@
+---
+title: "Working with React Server Components"
+slug: react-server-components
+authors:
+  - yunus
+tags: [redis, react, nextjs]
+---
+
+In this tutorial, we will understand React Server Components and with that knowledge, build a simple view counter for your app together. Feel free to jump to Project Setup if you want to go ahead with the implementation.
+
+## Understanding React Server Components
+
+In order to understand how React Server Components work, we will briefly look at Client-Side Rendering (CSR) and Server-Side Rendering (SSR).
+
+### Client-Side Rendering (CSR)
+
+In CSR, most of the rendering work happens in the browser on the client-side:
+
+1. User requests a site.
+2. Server sends a HTML file and links to CSS and JS files.
+3. Client downloads the JS resources.
+4. Client renders the page without content.
+5. Client fetches the data from the API in Server.
+6. Client re-renders the page with content.
+
+![Client-Side Rendering](/blog/react-server-components/csr.png)
+
+### Server-Side Rendering (SSR)
+
+In SSR, the server generates the full HTML content for each request and sends it to the client:
+
+1. User requests a site.
+2. Server renders the actual HTML files but they are not yet interactive.
+3. Client downloads the JS resources.
+4. In Client, using `hydrateRoot()` function, event listeners, state and other interactivity of React components are attached to the existing DOM elements generated by the Server, this is known as **hydration**.
+5. Client fetches the data from the API in Server.
+6. Client re-renders the page with content.
```

Comment:
I think the fetching happens before getting the initial html in ssr. The initial html you get is the actual final html, then the js gets downloaded for interactivity

Reply:
I thought of it like this, correct me if I am wrong:

Client Component Rendered with SSR:
```
...
export default function Views({ slug }: { slug: string }) {
  const [views, setViews] = useState<number>(0);

  useEffect(() => { // Get the number of views
    ...
  }, [slug]);
 ...
```
Final html is sent and hydrated as you mentioned. However useEffect is ran after the first render to invoke the API and get the data, as described in the blog.

Server Component:
```
...
export default async function Views({ slug }: { slug: string }) {
  const views = (await redis.get<number>(`views:${slug}`)) || 0; // Get the number of views
...
```
In this case, data fetching happens on the Server-Side before the final html is sent to the Client, as described in the blog.